API change: add initializationFailureThreshold to health probes

matthyx opened the pull request:
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**
Adds an numerical option `maxInitialFailureCount` to probes to allow a greater number of failures during the initial start of the pod before taking action.

**Which issue(s) this PR fixes**
Fixes #27114

**Special notes for your reviewer**
It's been a long time since I last contributed to k/k, I don't know where to document this change, please point me to the correct page.

**Does this PR introduce a user-facing change?**
Yes.
```release-note
Adds an numerical option `maxInitialFailureCount` to probes to allow a greater number of failures during the initial start of the pod before taking action.
```

matthyx committed:
Add initializationFailureThreshold to health probes

matthyx commented:
/assign @lavalamp

matthyx commented:
/test pull-kubernetes-local-e2e-containerized

matthyx commented:
/test pull-kubernetes-local-e2e-containerized

matthyx commented:
@lavalamp e2e-containerized test failure was a flake... you can review the PR.

lavalamp commented on pkg/kubelet/prober/worker.go:
```diff
@@ -218,6 +221,11 @@ func (w *worker) doProbe() (keepGoing bool) {
 	}
 
+	// Record the end of the initialization phase.
+	if result == results.Success && !w.hasInitialized {
```
We need to get someone from SIG Node to weigh in on this.

* Right now, this code might randomly allow MaxInitialFailureCount failures after the pod has become alive, if kubelet happens to restart at the same time the pod goes unhealthy. Is that OK from a UX perspective? From a Kubelet perspective? (i.e., do other things already behave like that?)

yujuhong replied:
Normally, kubelet handles container/pod lifecycle-related functions by relying on the underlying container runtime to persist the states. For probes and/or lifecycle hooks, kubelet rely on in-memory states only.

This is consistent with the existing probe behavior, but the existing behavior does not care about pod lifecycle much. If the MaxInitialFailureCount is set very high, it can potentially make the container downtime much longer and will need to be used with caution (documentation?).

By the way, just to make it clear, even though we use the term "pod" in the discussion threads, the changes pertain to "containers" and "container starts".

matthyx replied:
@yujuhong maybe we could persist the `hasInitialized` state somewhere inside the container itself? Why not a file that we create somewhere?
But it would do an exec every time we need to check that state...

matthyx replied:
> If the MaxInitialFailureCount is set very high, it can potentially make the container downtime much longer and will need to be used with caution (documentation?).

Indeed, but people in need of that feature would normally use a very high FailureThreshold, so it wouldn't be worse in that case...

matthyx replied:
Detailed answer here: https://github.com/kubernetes/enhancements/blob/4023cd9b30b33abe9b36b8f3826c13d02e92a450/keps/sig-node/20190221-maxInitialFailureCount-health-probes.md#stateless-kubelet

lavalamp commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1977,6 +1977,10 @@ type Probe struct {
 	// Defaults to 3. Minimum value is 1.
 	// +optional
 	FailureThreshold int32 `json:"failureThreshold,omitempty" protobuf:"varint,6,opt,name=failureThreshold"`
+	// Maximum number of failures allowed for the probe to be considered failed before having succeeded once.
```
We need to make it clearer that this is only while waiting for the pod to start up. We need to document the behavior when kubelet restarts (at a minimum); I was thinking it was OK to restart the count in that case but with fresh eyes it's not sounding like a good idea to me any more. :/

I'll suggest some wording in a bit.

We also need to document and/or enforce compatibility/version skew guarantees.

thockin commented on pkg/kubelet/prober/worker.go:
```diff
@@ -218,6 +221,11 @@ func (w *worker) doProbe() (keepGoing bool) {
 
+	// Record the end of the initialization phase.
+	if result == results.Success && !w.hasInitialized {
+		w.hasInitialized = true
```
if kubelet crashes and restarts, what happens?  Do we suddenly allow some (potentially large) number of failures again?

yujuhong replied:
Yes, based on the current implementation.

thockin reviewed (commented):
The main complexity of this overall idea is the state-saving, which I don't think is handled correctly here.  Node folks will know best.

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -1874,6 +1874,10 @@ type Probe struct {
 	// Minimum consecutive failures for the probe to be considered failed after having succeeded.
 	// +optional
 	FailureThreshold int32
+	// Maximum number of failures allowed for the probe to be considered failed before having succeeded once.
```
> We need to add some better documentation for this struct, I think, including a timeline example or 2 or 3 showing how the fields interact.

Agree

> If we're going to impose the requirement that kubelet carry this state at all (@dchen1107 @yujuhong @derekwaynecarr), it's the same burden in either mode.

FailureThreshold is already stateful and resets if the kubelet restarts... would MaxInitialFailureCount behave the same way (e.g. "initial" is from the probe's perspective, not the container's)?

yujuhong replied:
> FailureThreshold is already stateful and resets if the kubelet restarts... would MaxInitialFailureCount behave the same way (e.g. "initial" is from the probe's perspective, not the container's)?

+1. What's the expected behavior? If the state needs to be persisted over kubelet restarts, it'll need to be handled very differently from the existing probes. 

> Another alternative would be to add a timeout to PostStart handlers, and use that instead? E.g. Do whatever you need to do internal to your pod as PostStart. When you return from that, probes will begin. If the timeout is hit, it's the same as PostStart having failed. Presumably any state that Kubelet needs to track that is already implemented?

IIRC, kubelet doesn't save any state for running/completing poststart hooks. If we go this route, we'd still have to persist some state, and we'll have to teach the probers to wait until the hook is completed. Right now, the probers may start in parallel (i.e., triggered by running containers).

matthyx commented:
Then should we persist all these states in etcd? I am talking about FailureThreshold, MaxInitialFailureCount, ...

thockin commented:
storing in etcd (e.g. pod status) is possible, but a little awkward.  I
want to hear node team on this.

On Wed, Nov 28, 2018 at 1:08 PM Matthias Bertschy <notifications@github.com>
wrote:

> Then should we persist all these states in etcd? I am talking about
> FailureThreshold, MaxInitialFailureCount, ...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/71449#issuecomment-442605040>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVNdw6n9-gJ1fVrkvgdGOdSBPD_2pks5uzvtRgaJpZM4Y1INl>
> .
>

matthyx commented:
Any thoughts about storing these states (@dchen1107 @yujuhong @derekwaynecarr)?

thockin commented:
No sweat, no rudeness understood, just managing expectations. :)
On Fri, Nov 30, 2018 at 10:29 AM Matthias Bertschy
<notifications@github.com> wrote:
>
> Oops, sorry, I didn't want to sound impatient or rude.
> Happy Kubecon Seattle!
> See you hopefully in Barcelona next year (if I manage to convince my boss).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

matthyx commented:
@thockin @dchen1107 @yujuhong @derekwaynecarr, Kubecon is finished.
Can I get your thoughts about storing kubelet states in etcd (or somewhere else) to survive kubelet restarts?

matthyx commented:
Maybe a stateful kubelet could be the subject of a KEP, since this problem isn't unique to that PR.
Can we move forward on this PR and discuss state backup elsewhere?

matthyx commented:
/test pull-kubernetes-local-e2e-containerized
/test pull-kubernetes-e2e-kops-aws

matthyx commented:
/test pull-kubernetes-bazel-test

yujuhong commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1982,6 +1982,10 @@ type Probe struct {
 	// +optional
 	FailureThreshold int32 `json:"failureThreshold,omitempty" protobuf:"varint,6,opt,name=failureThreshold"`
+	// Maximum number of failures allowed for the probe to be considered failed before having succeeded once.
+	// No default value. Minimum value is 1.
```
Isn't the default value 0 in this case? This is larger than the minimum value already.

Also, there is no validation for the field.

yujuhong commented:
> storing in etcd (e.g. pod status) is possible, but a little awkward. I want to hear node team on this.

I don't think we should rely on etcd to persist the state. kubelet reports pod/container status, but does not rely on the apiserver to perform health check today. Changes like this may increase container downtime when kubelet the apisever is temporarily unreachable.

yujuhong commented:
Also, probably an unpopular opinion, but the 1.14 release process requires a KEP for features like this. I think the KEP deadline has passed, so this'd most likely miss the 1.14 release.
You're also welcome to add this in the sig-node weekly meeting agenda to get more attention: https://github.com/kubernetes/community/tree/master/sig-node

matthyx commented:
> this'd most likely miss the 1.14 release

It's OK, we've also missed the 1.13 deadline... at least it's moving a bit now, thanks for taking the time to give your point of view.

So you suggest I add this point to the sig-node meeting, would I have to participate to defend this idea?

yujuhong commented:
> So you suggest I add this point to the sig-node meeting, would I have to participate to defend this idea?

I wouldn't call it "defend", but you'll need to attend the meeting to explain what the proposal is about, and participate in the discussion.

matthyx commented:
> I wouldn't call it "defend", but you'll need to attend the meeting to explain what the proposal is about, and participate in the discussion.

Thanks for your help, I will shoot for the SIG meeting next week. In the meantime, I will try to sketch out the provisional KEP.

matthyx commented:
> Thanks for your help, I will shoot for the SIG meeting next week. In the meantime, I will try to sketch out the provisional KEP.

@yujuhong please have a look at the [KEP](https://github.com/matthyx/enhancements/blob/master/keps/sig-node/20190221-maxInitialFailureCount-health-probes.md), and could you schedule a discussion in an upcoming SIG-node meeting?

yujuhong commented:
> @yujuhong please have a look at the [KEP](https://github.com/matthyx/enhancements/blob/master/keps/sig-node/20190221-maxInitialFailureCount-health-probes.md), and could you schedule a discussion in an upcoming SIG-node meeting?

Added this to the 2/26 agenda. Any one who is a member of the sig-node mailing list can edit the doc directly.

matthyx commented:
> Added this to the 2/26 agenda. Any one who is a member of the sig-node mailing list can edit the doc directly.

I have created a PR and assigned sig-node: https://github.com/kubernetes/enhancements/pull/860

matthyx commented:
I have renamed the parameter (see KEP https://github.com/matthyx/enhancements/blob/master/keps/sig-node/20190221-initializationFailureThreshold-health-probes.md), added a default, fixed tests and rebased.

@lavalamp I think you can review again if you want, and also tell me about the KEP, pls.

lavalamp commented:
Sorry, I just saw that it was my turn on this. Is the KEP merged somewhere?

lavalamp commented:
Ah, I found the KEP PR. You may need to reach out to people on slack--github notifications tend to get lost in the noise :/

matthyx commented:
/test pull-kubernetes-bazel-test

matthyx commented:
This is superceded by startupProbe described in https://github.com/kubernetes/enhancements/issues/950

matthyx commented:
Superseded by https://github.com/kubernetes/kubernetes/pull/77807
